core/src/prediction/LLCausalPredictionStrategy.cpp

```diff
@@ -147,8 +150,152 @@ std::vector<double> LLCausalPredictionStrategy::compute_variance(
         size_t sampleID,
         std::vector<std::vector<size_t>> samples_by_tree,
         std::unordered_map<size_t, double> weights_by_sampleID,
-        const Data* train_data,
-        const Data* data,
+        const Data* original_data,
+        const Data* test_data,
         size_t ci_group_size){
-  return { 0.0 };
+
+  double lambda = lambdas[0];
+
+  size_t num_variables = linear_correction_variables.size();
+  size_t num_nonzero_weights = weights_by_sampleID.size();
+
+  std::vector<size_t> sample_index_map(original_data->get_num_rows());
+  std::vector<size_t> indices(num_nonzero_weights);
+
+  Eigen::MatrixXd weights_vec = Eigen::VectorXd::Zero(num_nonzero_weights);
+  {
+    size_t i = 0;
+    for (auto& it : weights_by_sampleID) {
+      size_t index = it.first;
+      double weight = it.second;
+      indices[i] = index;
+      sample_index_map[index] = i;
+      weights_vec(i) = weight;
+      i++;
+    }
+  }
+
+  size_t dim_X = 2 * num_variables + 2;
+  Eigen::MatrixXd X (num_nonzero_weights, dim_X);
+  Eigen::MatrixXd Y (num_nonzero_weights, 1);
+  size_t treatment_index = num_variables + 1;
+
+  for (size_t i = 0; i < num_nonzero_weights; ++i) {
+    // Index of next neighbor with nonzero weights
+
+    size_t index = indices[i];
+    double treatment = original_data->get_treatment(index);
+
+    // Intercept
+    X(i, 0) = 1;
+    //X_alone(i, 0) = 1;
+
+    // Regressors
+    for (size_t j = 0; j < num_variables; ++j){
+      size_t current_predictor = linear_correction_variables[j];
+      // X - x0 column
+      X(i,j+1) = original_data->get(index, current_predictor) -
+                 test_data->get(sampleID, current_predictor);
+
+      // (X - x0)*W column
+      X(i, treatment_index + j + 1) = X(i, j+1) * treatment;
+    }
+
+    // Treatment (just copied)
+    X(i, treatment_index) = treatment;
+
+    // Outcome (just copied)
+    Y(i) = original_data->get_outcome(index);
+  }
+
+  // find ridge regression predictions
+  Eigen::MatrixXd M_unpenalized (dim_X, dim_X);
+  M_unpenalized.noalias() = X.transpose() * weights_vec.asDiagonal() * X;
+  double normalization = M_unpenalized.trace() / dim_X;
```

Comment:
Small comment, we could move this into the `if (!weight_penalty) { ... }` block to avoid calculating the normalization when using a weighted penalty.